### PR TITLE
[CBR] implements all the APIs of Tag.

### LINKED ISSUE #370: [CBR]: Add tag management for vaults
Tags management API for CBR vault is implemented in OTC.

Please add this functionality for gophertelekomcloud.

The API that needs to be implemented:
Batch adding of tags
Batch deleting of tags
Query tags of an associated resource

### PR DESCRIPTION
**Pull request by Aloento**
### What this PR does / why we need it

The customer wants to have the possibility to update CBR Vault `tags`.
https://jira.tsi-dev.otc-service.com/browse/DM-3663

### Which issue this PR fixes

fixes #370

### Special notes for your reviewer

I cannot write the corresponding test
you can either implement it or wait for my supervisor to arrange an available OTC environment for me.

### CODE REVIEW

File: openstack/cbr/v3/tags/requests.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	// Returns the full amount of data when there are no filter conditions
+	Tags []Tag `json:"tags,omitempty"`
+	// Include any tag
+	TagsAny []Tag `json:"tags_any,omitempty"`
+	// Label not included
+	NotTags []Tag `json:"not_tags,omitempty"`
+	// Does not contain any tags
+	NotTagsAny []Tag `json:"not_tags_any,omitempty"`
+	// Only the op_service permission can use this field as a resource instance filter condition
+	SysTags []Tag `json:"sys_tags,omitempty"`
```
Comment (anton-sidelnikov) [inline openstack/cbr/v3/tags/requests.go]: if this resources not persisted in response, please remove or comment

File: openstack/cbr/v3/tags/requests.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+)
+
+/*
+ShowVaultProjectTag
+Query the set of all tags of the tenant in the specified Region and instance type
+
+@author Aloento
+@since 0.5.17
+@version 0.1.0
+*/
```
Comment (anton-sidelnikov) [inline openstack/cbr/v3/tags/requests.go]: @gtema should we keep this lines or better remove? We never added smth like this before.

**Review by artem-lifshits — CHANGES_REQUESTED** [review]
Please add acceptance tests.

Provided documentation is not final, for example "querying vault resources" API is not registered, other API depending on enterprise_project also have some unnecessary parameters (like sys_tags).

File: openstack/cbr/v3/tags/requests.go (hunk 1 above)
Comment (Aloento) [reply openstack/cbr/v3/tags/requests.go]: please check `Table 2-290` in `Cloud Backup and Recovery API Reference_18072022_alifshit`

**Comment by Aloento** [comment]
@artem-lifshits I am not familiar with the test of our project, so I may have some problems with the test I committed, please help me to check.

File: acceptance/openstack/cbr/v3/tags_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/opentelekomcloud/gophertelekomcloud/acceptance/clients"
+	"github.com/opentelekomcloud/gophertelekomcloud/openstack/cbr/v3/tags"
+	th "github.com/opentelekomcloud/gophertelekomcloud/testhelper"
+)
+
+func TestTags(t *testing.T) {
+	client, err := clients.NewCbrV3Client()
+	th.AssertNoErr(t, err)
+
+	tagsResp := tags.ShowVaultProjectTag(client)
```
Comment (anton-sidelnikov) [inline acceptance/openstack/cbr/v3/tags_test.go]: Please use more descriptive names, here you can use projectTags, for example

Comment (anton-sidelnikov) [reply acceptance/openstack/cbr/v3/tags_test.go]: do .Extract() here

File: acceptance/openstack/cbr/v3/tags_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+		Key:   tag.Key,
+		Value: tag.Values[0],
+	}
+
+	req := tags.ResourceInstancesRequest{
+		Tags:   []tags.Tag{tag},
+		Action: tags.Filter,
+	}
+
+	insResp := tags.ShowVaultResourceInstances(client, req)
```
Comment (anton-sidelnikov) [inline acceptance/openstack/cbr/v3/tags_test.go]: variable name and extract here

Comment (anton-sidelnikov) [reply acceptance/openstack/cbr/v3/tags_test.go]: same mistakes below

File: acceptance/openstack/cbr/v3/tags_test.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+	vaultResp := tags.ShowVaultTag(client, resId)
+	vaultRes, err := vaultResp.Extract()
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, len(vaultRes.Tags) > 0, true)
+
+	delResp := tags.DeleteVaultTag(client, resId, tag.Key)
+	th.AssertNoErr(t, delResp.ExtractErr())
+
+	createResp := tags.CreateVaultTags(client, resId, sysTag)
```
Comment (anton-sidelnikov) [inline acceptance/openstack/cbr/v3/tags_test.go]: could you check some actual value and key of tag?

File: openstack/cbr/v3/tags/requests.go (hunk 2 above)
Comment (anton-sidelnikov) [inline openstack/cbr/v3/tags/requests.go]: i think code no need to be signed, you already in contributors

File: openstack/cbr/v3/tags/requests.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	// Returns the full amount of data when there are no filter conditions
+	Tags []Tag `json:"tags,omitempty"`
+	// Backups with any tags in this list will be filtered.
+	TagsAny []Tag `json:"tags_any,omitempty"`
+	// Backups without these tags will be filtered.
+	NotTags []Tag `json:"not_tags,omitempty"`
+	// Backups without any tags in this list will be filtered.
+	NotTagsAny []Tag `json:"not_tags_any,omitempty"`
+	// Only the op_service permission can use this field as a resource instance filter condition
+	SysTags []Tag `json:"sys_tags,omitempty"`
```
Comment (anton-sidelnikov) [inline openstack/cbr/v3/tags/requests.go]: Please check which tags cannot be set, and remove it from schema

File: openstack/cbr/v3/tags/results.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	Tags           []SysTag `json:"tags"`
+	ResourceName   string   `json:"resource_name"`
+	SysTags        []SysTag `json:"sys_tags"`
+}
+
+type Vault struct {
+	vaults.Vault
+	SmnNotify bool `json:"smn_notify"`
+	Threshold int  `json:"threshold"`
+}
```
Comment (anton-sidelnikov) [inline openstack/cbr/v3/tags/results.go]: strange structure, do you need it here?

**Review by anton-sidelnikov — CHANGES_REQUESTED** [review]

File: openstack/cbr/v3/tags/requests.go (hunk 2 above)
Comment (gtema) [reply openstack/cbr/v3/tags/requests.go]: pls drop. We may add a general copyright header though (smth like # Copyright 2022 Open Telekom Cloud, T-Systems International GmbH), but that does not need to be done now.

File: openstack/cbr/v3/tags/results.go (hunk 7 above)
Comment (Aloento) [reply openstack/cbr/v3/tags/results.go]: We do not support these two fields and have removed them.

File: openstack/cbr/v3/tags/requests.go (hunk 6 above)
Comment (Aloento) [reply openstack/cbr/v3/tags/requests.go]: We do not support SysTags and have removed them.

File: openstack/cbr/v3/tags/requests.go (hunk 2 above)
Comment (Aloento) [reply openstack/cbr/v3/tags/requests.go]: Removed.

File: acceptance/openstack/cbr/v3/tags_test.go (hunk 3 above)
Comment (Aloento) [reply acceptance/openstack/cbr/v3/tags_test.go]: done

File: acceptance/openstack/cbr/v3/tags_test.go (hunk 4 above)
Comment (Aloento) [reply acceptance/openstack/cbr/v3/tags_test.go]: done

File: openstack/cbr/v3/tags/urls.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
+/*
+@author Aloento
+@since 0.4.17
+@version 0.1.0
+*/
```
Comment (anton-sidelnikov) [inline openstack/cbr/v3/tags/urls.go]: here also

**Review by anton-sidelnikov — CHANGES_REQUESTED** [review]

File: acceptance/openstack/cbr/v3/tags_test.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+	resourceID := instances.Resources[0].ResourceID
+
+	vaultTags, err := tags.ShowVaultTag(client, resourceID).Extract()
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, len(vaultTags.Tags) > 0, true)
+
+	th.AssertNoErr(t, tags.DeleteVaultTag(client, resourceID, testTag.Key).ExtractErr())
+
+	th.AssertNoErr(t, tags.CreateVaultTags(client, resourceID, monoTag).ExtractErr())
```
Comment (anton-sidelnikov) [inline acceptance/openstack/cbr/v3/tags_test.go:71]: i think here is better to manually create some tag, like: 
  ```
tags.MonoTag{
		  Key:   "some_key",
		  Value: "some_value,
	  }
```
and then assert result of creation

**Review by anton-sidelnikov — CHANGES_REQUESTED** [review]

File: openstack/cbr/v3/tags/urls.go (hunk 8 above)
Comment (Aloento) [reply openstack/cbr/v3/tags/urls.go]: done

File: acceptance/openstack/cbr/v3/tags_test.go (hunk 9 above)
Comment (Aloento) [reply acceptance/openstack/cbr/v3/tags_test.go:71]: done

File: acceptance/openstack/cbr/v3/tags_test.go (hunk 5 above)
Comment (Aloento) [reply acceptance/openstack/cbr/v3/tags_test.go]: see below

File: acceptance/openstack/cbr/v3/tags_test.go (hunk 9 above)
Comment (anton-sidelnikov) [reply acceptance/openstack/cbr/v3/tags_test.go:71]: mm, nothing changed?

**Review by anton-sidelnikov — DISMISSED** [review]

**Review by artem-lifshits — DISMISSED** [review]

**Review by artem-lifshits — DISMISSED** [review]

**Comment by anton-sidelnikov** [comment]
tags_test.go:54: Failure in tags_test.go, line 54: unexpected error "failed to extract Vault Resource Instances"

**Comment by Aloento** [comment]
=== RUN   TestTags
--- PASS: TestTags (3.47s)
PASS

File: results.go [hunk 10]
```diff
@@ -244,6 +244,31 @@ func (r ErrResult) ExtractErr() error {
...
+}
+
+func (r Result) Extract() (*ErrRespond, error) {
+	var s = ErrRespond{}
+	if r.Err != nil {
+		return nil, r.Err
+	}
+	err := r.ExtractInto(&s)
+	if err != nil {
+		return nil, fmt.Errorf("failed to ErrWithResult")
```
Comment (anton-sidelnikov) [inline results.go]: missing format?

**Review by anton-sidelnikov — CHANGES_REQUESTED** [review]

**Review by anton-sidelnikov — APPROVED** [review]

**Review by artem-lifshits — APPROVED** [review]

**Pull request merged**